accounts/abi/unpack.go

```diff
@@ -33,43 +34,72 @@ var (
 )
 
 // ReadInteger reads the integer based on its kind and returns the appropriate value.
-func ReadInteger(typ Type, b []byte) interface{} {
+func ReadInteger(typ Type, b []byte) (interface{}, error) {
+	ret := new(big.Int).SetBytes(b)
+
 	if typ.T == UintTy {
+		u64, isu64 := ret.Uint64(), ret.IsUint64()
 		switch typ.Size {
 		case 8:
-			return b[len(b)-1]
+			if !isu64 || u64 > math.MaxUint8 {
+				return nil, errBadUint8
+			}
+			return byte(u64), nil
 		case 16:
-			return binary.BigEndian.Uint16(b[len(b)-2:])
+			if !isu64 || u64 > math.MaxUint16 {
+				return nil, errBadUint16
+			}
+			return uint16(u64), nil
 		case 32:
-			return binary.BigEndian.Uint32(b[len(b)-4:])
+			if !isu64 || u64 > math.MaxUint32 {
+				return nil, errBadUint32
+			}
+			return uint32(u64), nil
 		case 64:
-			return binary.BigEndian.Uint64(b[len(b)-8:])
+			if !isu64 {
+				return nil, errBadUint64
+			}
+			return u64, nil
 		default:
 			// the only case left for unsigned integer is uint256.
-			return new(big.Int).SetBytes(b)
+			return ret, nil
 		}
 	}
+
+	// big.SetBytes can't tell if a number is negative or positive in itself.
+	// On EVM, if the returned number > max int256, it is negative.
+	// A number is > max int256 if the bit at position 255 is set.
+	if ret.Bit(255) == 1 {
+		ret.Add(MaxUint256, new(big.Int).Neg(ret))
+		ret.Add(ret, common.Big1)
+		ret.Neg(ret)
+	}
+	i64, isi64 := ret.Int64(), ret.IsInt64()
 	switch typ.Size {
 	case 8:
-		return int8(b[len(b)-1])
+		if !isi64 || i64 < math.MinInt8 || i64 > math.MaxInt8 {
+			return nil, errBadInt8
+		}
+		return int8(i64), nil
```

Comment:
This does not seem to be covered by the testcases in `packing_test.go` -- since nothing breaks. BUT: It looks to me like you are changing the behaviour for negative numbers here? 
I'll look at bit closer at this later on

Reply:
Without your PR, these cases work: 
```
	{
		def:      `[{"type": "int8"}]`,
		unpacked: int8(-1),
		packed:   "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
	},
	{
		def:      `[{"type": "int8"}]`,
		unpacked: int8(-1),
		packed:   "00000000000000000000000000000000000000000000000000000000000000ff",
	},
```

Reply:
WIth your changes, this is a case which fails, and which should not fail (?)
```
    unpack_test.go:50: test 15 ([{"type": "int8"}]: 00000000000000000000000000000000000000000000000000000000000000ff) failed: abi: improperly encoded int8 value
```
more precisely:
```
	{
		def:      `[{"type": "int8"}]`,
		unpacked: int8(-1),
		packed:   "00000000000000000000000000000000000000000000000000000000000000ff",
	},
```

Reply:
Also, this seems to be valid way to encode `int8` even after the changes in this PR: 
```
	{
		def:      `[{"type": "int8"}]`,
		unpacked: int8(-1),
		packed:   "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
	},
```

Reply:
The question is: will EVM represents a int8(-1) as `00000000000000000000000000000000000000000000000000000000000000ff ` in the first place? It feels like an invalid case.

Reply:
No, `int8(-1)` is represented as all `f`. In EVM, a signed number is negative when the most significant bit is `1`.

Reply:
Yeah I seem to have been wrong